lib/handler/mruby.c

```diff
@@ -265,6 +269,64 @@ static mrb_value build_constants(mrb_state *mrb, const char *server_name, size_t
     return ary;
 }
 
+static void handle_exception(h2o_mruby_context_t *ctx, h2o_mruby_generator_t *generator)
+{
+    mrb_state *mrb = ctx->shared->mrb;
+
+    if (generator == NULL) {
+        fprintf(stderr, "mruby raised: %s\n", RSTRING_PTR(mrb_inspect(mrb, mrb_obj_value(mrb->exc))));
+    } else {
+        assert(generator->req != NULL);
+        h2o_req_log_error(generator->req, H2O_MRUBY_MODULE_NAME, "mruby raised: %s\n", RSTRING_PTR(mrb_inspect(mrb, mrb_obj_value(mrb->exc))));
+        if (generator->req->_generator == NULL) {
+            h2o_send_error_500(generator->req, "Internal Server Error", "Internal Server Error", 0);
+        } else {
+            h2o_mruby_send_chunked_close(generator);
+        }
+    }
+    mrb->exc = NULL;
+}
+
+mrb_value send_error_callback(h2o_mruby_context_t *ctx, mrb_value input, mrb_value receiver, mrb_value args, int *run_again)
+{
+    mrb_state *mrb = ctx->shared->mrb;
+    mrb->exc = mrb_obj_ptr(mrb_ary_entry(args, 0));
+    h2o_mruby_generator_t *generator = h2o_mruby_get_generator(mrb, mrb_ary_entry(args, 1));
+    handle_exception(ctx, generator);
+    return mrb_nil_value();
+}
+
+mrb_value pend_request_callback(h2o_mruby_context_t *ctx, mrb_value input, mrb_value receiver, mrb_value args, int *run_again)
+{
+    mrb_state *mrb = ctx->shared->mrb;
+    mrb_value pending = mrb_ary_new_capa(mrb, 2);
+    mrb_ary_set(mrb, pending, 0, ctx->proc);
+    mrb_ary_set(mrb, pending, 1, input);
+    mrb_ary_push(mrb, ctx->pendings, pending);
+    return mrb_nil_value();
+}
+
+mrb_value process_pending_requests_callback(h2o_mruby_context_t *ctx, mrb_value input, mrb_value receiver, mrb_value args, int *run_again)
+{
+    mrb_state *mrb = ctx->shared->mrb;
+    mrb_int i;
+    mrb_int len = RARRAY_LEN(ctx->pendings);
+    for (i = 0; i != len; ++i) {
+        mrb_value pending = mrb_ary_entry(ctx->pendings, i);
+        mrb_value pending_resumer = mrb_ary_entry(pending, 0);
+        mrb_value pending_input = mrb_ary_entry(pending, 1);
+        h2o_mruby_run_fiber(ctx, pending_resumer, pending_input, NULL);
+    }
+    ctx->pendings = mrb_nil_value();
+
+    mrb_value exc = mrb_ary_entry(args, 0);
+    if (! mrb_nil_p(exc)) {
+        mrb->exc = mrb_obj_ptr(exc);
+        handle_exception(ctx, NULL);
+    }
```

Comment:
Would it make more sense to report the exception (by calling `handle_exception`) then execute the pending requests?

Reply:
Hm? Why do you think so?

Reply:
Because users would expect to see the exceptions reported in the order they happened.

`exc` found in `args` is an error that has occurred prior to those that might be reported while we call the run_fiber function just above this code. Hence we have a possibility of reporting errors in the wrong order.

Reply:
あぁ、たしかし。:thumbsup: